File: Introduction to Problem Solving I/03. Introduction to Arrays/In Class/JavaScript/rotateArray.js
// # Rotating Array in JavScript 🤌🏻

const reverseArray = function (startIndex, endIndex, arr) {
  while (startIndex < endIndex) {
    const temp = arr[startIndex];
    arr[startIndex] = arr[endIndex];
    arr[endIndex] = temp;

    startIndex++;
    endIndex--;
  }
};

const rotateArray = function (K, arr) {
  const N = arr.length;
  K = K % N;

  // * firstly reversing entire array
  reverseArray(0, N - 1, arr);

  // * reverse first half of array
  reverseArray(0, K - 1, arr);

  // * reverse second half of array
  reverseArray(K, N - 1, arr);
};

const arr = [1, 2, 3, 4, 5, 6, 7];
rotateArray(4, arr);
console.log("ater shifting: " + arr);

File: 02. Introduction to Problem Solving I/03. Introduction to Arrays/In Class/Python/rotateArray.py
# rotating array with python 🐍

def reverse_list(start_index: int, end_index: int, arr: list):
    while start_index < end_index:
        temp = arr[start_index]
        arr[start_index] = arr[end_index]
        arr[end_index] = temp

        start_index += 1
        end_index -= 1

def rotate_list(K: int, arr: list):
    N = len(arr)
    K = K % N

    # reversing whole
    reverse_list(0, N - 1, arr)

    # reversing first half
    reverse_list(0, K - 1, arr)

    # reversing second half
    reverse_list(K, N - 1, arr)

arr = [1, 2, 3, 4, 5, 6, 7]
rotate_list(K=4, arr=arr)
print(f"after rotating 4 times: {arr}")